honeycombio/data_source_column.go

```diff
@@ -0,0 +1,102 @@
+package honeycombio
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	honeycombio "github.com/honeycombio/terraform-provider-honeycombio/client"
+)
+
+func dataSourceHoneycombioColumn() *schema.Resource {
+	return &schema.Resource{
+		ReadContext: dataSourceHoneycombioColumnRead,
+
+		Description: `
+'honeycombio_column' data source provides details about a specific column in a dataset, matched by name.
+`,
+
+		Schema: map[string]*schema.Schema{
+			"dataset": {
+				Type:        schema.TypeString,
+				Optional:    false,
+				Required:    true,
+				Description: "The dataset this column is associated with",
+			},
+			"name": {
+				Type:        schema.TypeString,
+				Optional:    false,
+				Required:    true,
+				Description: "Name of the column",
+			},
+			// Generated attributes
+			"hidden": {
+				Type:        schema.TypeBool,
+				Required:    false,
+				Optional:    false,
+				Computed:    true,
+				Description: "Whether the column is hidden",
+			},
+			"type": {
+				Type:        schema.TypeString,
+				Required:    false,
+				Optional:    false,
+				Computed:    true,
+				Description: "The type of column, allowed types are `string`, `integer`, `float`, and `boolean`.",
+			},
+			"description": {
+				Type:        schema.TypeString,
+				Required:    false,
+				Optional:    false,
+				Computed:    true,
+				Description: "A description of the column",
+			},
+			"last_written": {
+				Type:        schema.TypeString,
+				Required:    false,
+				Optional:    false,
+				Computed:    true,
+				Description: "The last time the column was written to",
+			},
+			"created_at": {
+				Type:        schema.TypeString,
+				Required:    false,
+				Optional:    false,
+				Computed:    true,
+				Description: "The time the column was created",
+			},
+			"updated_at": {
+				Type:        schema.TypeString,
+				Required:    false,
+				Optional:    false,
+				Computed:    true,
+				Description: "The time the column was last updated",
+			},
+		},
+	}
+}
+
+func dataSourceHoneycombioColumnRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*honeycombio.Client)
+
+	dataset := d.Get("dataset").(string)
+	matchName := d.Get("name").(string)
+
+	column, err := client.Columns.GetByKeyName(ctx, dataset, matchName)
+	if err != nil {
+		return diag.FromErr(fmt.Errorf("column not found, API returned: %s", err.Error()))
```

Comment:
I'm wondering if this is the idiomatic behavior here or not?   Currently this makes TF bomb out if you try to get data on a column which doesn't exist - is that right, or would it be better to return _something_?

Reply:
There are both approaches out there in the ecosystem.

So far in this provider the plural version of a datasource will happily return an empty list (no failure), but the singular fails if you don't return a match.

I think this makes sense given that you're asking for a specific thing with the singular and if you don't get it the only behaviour that seems to make sense is to raise an error (e.g. "I failed to lookup what you asked me for").

That said, I'm open to alternatives :) What's the use case for returning nothing here and potentially having weird errors when trying to access the datasource?

Reply:
You're right,  after thinking it through some more I'm with you